src/promisify.js

```diff
@@ -0,0 +1,132 @@
+/***************************************************************************************************
+
+The promisify implementation found in this file is derived from the implementation in Node.js. Its
+license (https://github.com/nodejs/node/blob/master/LICENSE) is as follows:
+
+Copyright Joyent, Inc. and other Node contributors. All rights reserved. Permission is hereby
+granted, free of charge, to any person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the Software without restriction, including without
+limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial
+portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+***************************************************************************************************/
+
+
+const kCustomPromisifiedSymbol = Symbol.for('util.promisify.custom');
+const callbackBuilder = require('./callbackBuilder');
+
+/**
+ * Promisify the given function.
+ *
+ * @param {Function} orig The function to promisify.
+ * @param {Boolean|String[]} options.variadic The variadic option informs how promisify will handle
+ *   more than one value - by default, promisify will only resolve the promise with the first value.
+ *     false    - only resolve the promise with the first value, default behavior
+ *     true     - resolve the promise with an array containing the variadic arguments
+ *     String[] - the names of the arguments to the callback, which will be used to create an object
+ *       of values.
+ * @return {Function: Promise} The promisified function.
+ */
+function promisify(orig, options = {}) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('promisify requires a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    if (typeof fn !== 'function') {
+      throw new TypeError('The [util.promisify.custom] property must be a function');
+    }
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  function fn(...args) {
+    return new Promise((resolve, reject) => {
+      try {
+        orig.call(this, ...args, callbackBuilder(resolve, reject, options));
```

Comment:
@wearhere I think maybe I should call `callbackBuilder` outside the `try`-`catch`, so just in case some error is generated because someone is doing something strange, it doesn't end up inside the promise.